cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java

```diff
@@ -71,12 +83,47 @@ public RuntimeInfo run(Program program, ProgramOptions options) {
     ProgramRunner runner = programRunnerFactory.create(ProgramRunnerFactory.Type.valueOf(program.getType().name()));
     Preconditions.checkNotNull(runner, "Fail to get ProgramRunner for type " + program.getType());
     ProgramOptions optionsWithRunId = addRunId(options, RunIds.generate());
-    final RuntimeInfo runtimeInfo = createRuntimeInfo(runner.run(program, optionsWithRunId), program);
+    ProgramOptions optionsWithPlugins = addPluginLocations(optionsWithRunId,
+                                                           Id.Namespace.from(program.getNamespaceId()),
+                                                           program.getApplicationSpecification());
+    final RuntimeInfo runtimeInfo = createRuntimeInfo(runner.run(program, optionsWithPlugins), program);
     programStarted(runtimeInfo);
     return runtimeInfo;
   }
 
   /**
+   * Return the copy of the {@link ProgramOptions} including locations of plugin artifacts in it.
+   * @param options the {@link ProgramOptions} in which the locations of plugin artifacts needs to be included
+   * @param namespace namespace of the Program
+   * @param appSpec program's Application Specification
+   * @return the copy of the program options with locations of plugin artifacts included in them
+   */
+  private ProgramOptions addPluginLocations(ProgramOptions options, Id.Namespace namespace,
+                                            ApplicationSpecification appSpec) {
+    if (appSpec == null) {
+      return options;
+    }
+
+    try {
+      Id.Artifact appArtifactId = Id.Artifact.from(namespace, appSpec.getArtifactId());
+      ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+      builder.putAll(options.getArguments().asMap());
+      Map<String, String> artifactLocations = Maps.newHashMap();
+      for (Map.Entry<String, Plugin> pluginEntry : appSpec.getPlugins().entrySet()) {
+        Plugin plugin = pluginEntry.getValue();
+        Map.Entry<ArtifactDescriptor, PluginClass> artifactEntry = artifactRepository.getPlugin(
+          appArtifactId, plugin.getPluginClass().getType(), plugin.getPluginClass().getName(), plugin.getArtifactId());
+        artifactLocations.put(plugin.getArtifactId().toString(), artifactEntry.getKey().getLocation().getName());
+      }
+      builder.put(ProgramOptionConstants.PLUGIN_FILENAMES, GSON.toJson(artifactLocations));
+      return new SimpleProgramOptions(options.getName(), new BasicArguments(builder.build()),
+                                      options.getUserArguments(), options.isDebug());
+    } catch (PluginNotExistsException | IOException e) {
+      return options;
```

Comment:
though I'm also confused why we need to add the filenames as an option.  Shouldn't we be able to tell the exact path based on the plugin artifact's id?  In fact it is a bug to use the actual filename here since it opens this up to race conditions.  What if the filename changes between now and whenever you pull the files down from hdfs?